src/material-experimental/mat-feature-highlight/feature-highlight-container.ts

```diff
@@ -0,0 +1,533 @@
+        radius = this._getDistance(this._targetCenter, {
+          x: calloutLeft,
+          y: calloutTop,
+        });
+        break;
+      case 'top_end':
+        radius = this._getDistance(this._targetCenter, {
+          x: calloutLeft + calloutWidth,
+          y: calloutTop,
+        });
+        break;
+      case 'bottom_start':
+        radius = this._getDistance(this._targetCenter, {
+          x: calloutLeft,
+          y: calloutTop + calloutHeight,
+        });
+        break;
+      case 'bottom_end':
+        radius = this._getDistance(this._targetCenter, {
+          x: calloutLeft + calloutWidth,
+          y: calloutTop + calloutHeight,
+        });
+        break;
+      default:
+        throw new Error(
+            `Cannot handle callout position ${this.config.calloutPosition}.`);
+    }
+
+    return `${Math.floor(radius * 2)}px`;
+  }
+
+  /**
+   * Position the root div element so that it's overlayed with the target
+   * element.
+   */
+  private _positionRootDivElement() {
+    // Remove the left and top first so that the offsetLeft and offsetTop can
+    // be computed correctly.
+    const style = this.rootDiv.nativeElement.style;
+    this._updateLeft(style, undefined);
+    this._updateTop(style, undefined);
+
+    // Only need to position the root element when the outer circle is bounded,
+    // i.e. overlay is not needed.
+    if (!this.config.isOuterCircleBounded) {
+      return;
+    }
+
+    const targetOffsetLeft = this._getOffsetLeft(this._getTargetElement());
+    const containerOffsetLeft =
+        this._getOffsetLeft(this._elementRef.nativeElement);
+    const left = targetOffsetLeft - containerOffsetLeft;
+    const top = this._getTargetElement().offsetTop -
+        this._elementRef.nativeElement.offsetTop;
+
+    this._updateLeft(style, left);
+    this._updateTop(style, top);
+  }
+
+  /** Activate feature highlight and display the animation. */
+  private _activate() {
+    this.layout();
+
+    if (this.state === 'active_with_animation') {
+      return;
+    }
+
+    this._changeAnimationState('active_without_animation');
+    this._addEventListeners();
+
+    // TODO(b/138400499) - Ideally we should emit afterOpened event when the
+    // transitionend is triggered. However it doesn't look like catalyst flush()
+    // triggers the transitionend event so there is no way to test the behavior.
+    // We may be able to use flushMicrotasks() when migrating the component to
+    // Angular upstream.
+    setTimeout(() => {
+      this._changeAnimationState('active_with_animation');
+      this.afterOpened.emit();
+    }, OPEN_ANIMATION_TIME_MS);
+
+    // Mark the container for check so that it can react if the view container
+    // is usign OnPush change detection.
+    this._changeDetectorRef.markForCheck();
+  }
+
+  private _changeAnimationState(newState: State) {
+    requestAnimationFrame(() => {
+      this.state = newState;
+      this._changeDetectorRef.markForCheck();
```

Comment:
I think this is redundant since `requestAnimationFrame` triggers change detection.